handlers.c

```diff
@@ -450,44 +451,45 @@ bool handler__list(globals_t *vars, char **argv, unsigned argc)
     return true;
 }
 
-/* XXX: handle multiple deletes, eg delete !1 2 3 4 5 6 */
 bool handler__delete(globals_t * vars, char **argv, unsigned argc)
 {
-    unsigned id;
-    char *end = NULL;
+    unsigned       id;
     match_location loc;
+    struct set     del_set;
 
     if (argc != 2) {
         show_error("was expecting one argument, see `help delete`.\n");
         return false;
     }
 
-    /* parse argument */
-    id = strtoul(argv[1], &end, 0x00);
-
-    /* check that strtoul() worked */
-    if (argv[1][0] == '\0' || *end != '\0') {
-        show_error("sorry, couldnt parse `%s`, try `help delete`\n", argv[1]);
+    if (vars->num_matches == 0) {
+        show_error("nothing to delete.\n");
         return false;
     }
-    
-    loc = nth_match(vars->matches, id);
-    
-    if (loc.swath)
-    {
-        /* it is not convenient to check whether anything else relies on this,
-           so just mark it as not a REAL match */
-        zero_match_flags(&loc.swath->data[loc.index].match_info);
-        vars->num_matches--;
-        return true;
-    }
-    else
-    {
-        /* I guess this is not a valid match-id */
-        show_warn("you specified a non-existant match `%u`.\n", id);
-        show_info("use \"list\" to list matches, or \"help\" for other commands.\n");
+
+    if (!parse_uintset(argv[1], &del_set, (size_t)vars->num_matches)) {
+        show_error("failed to parse the set, try `help delete`\n");
         return false;
     }
+
+    foreach_set_bw(i, &del_set) {
+        id  = ((unsigned *)del_set.buf)[i];
```

Comment:
There is a nasty bug here.

```
14> list
[ 0] 559d0632e41a,  1 +          41a,  heap, 2, [I32 ]
[ 1] 7f22b0bb8300,  3 +         3300,  misc, 2, [I32 ]
[ 2] 7f22b0dde0a2,  6 +           a2,  misc, 2, [I32 ]
[ 3] 7f22b0dde120,  6 +          120,  misc, 2, [I32 ]
[ 4] 7f22b0dde45c,  6 +          45c,  misc, 2, [I32 ]
[ 5] 7f22b0dde570,  6 +          570,  misc, 2, [I32 ]
[ 6] 7f22b0dde5d1,  6 +          5d1,  misc, 2, [I32 ]
[ 7] 7f22b0ddeae0,  6 +          ae0,  misc, 2, [I32 ]
[ 8] 7ffd4cde3bb0,  7 +        1ebb0, stack, 2, [I32 ]
[ 9] 7ffd4cde3eb3,  7 +        1eeb3, stack, 2, [I32 ]
[10] 7ffd4cde3f7a,  7 +        1ef7a, stack, 2, [I32 ]
[11] 7ffd4cde4010,  7 +        1f010, stack, 2, [I32 ]
[12] 7ffd4cde4854,  7 +        1f854, stack, 2, [I32 ]
[13] 7ffd4cde4910,  7 +        1f910, stack, 2, [I32 ]
14> delete 1,2
12> list
[ 0] 7f22b0bb8300,  3 +         3300,  misc, 2, [I32 ]
[ 1] 7f22b0dde120,  6 +          120,  misc, 2, [I32 ]
[ 2] 7f22b0dde45c,  6 +          45c,  misc, 2, [I32 ]
[ 3] 7f22b0dde570,  6 +          570,  misc, 2, [I32 ]
[ 4] 7f22b0dde5d1,  6 +          5d1,  misc, 2, [I32 ]
[ 5] 7f22b0ddeae0,  6 +          ae0,  misc, 2, [I32 ]
[ 6] 7ffd4cde3bb0,  7 +        1ebb0, stack, 2, [I32 ]
[ 7] 7ffd4cde3eb3,  7 +        1eeb3, stack, 2, [I32 ]
[ 8] 7ffd4cde3f7a,  7 +        1ef7a, stack, 2, [I32 ]
[ 9] 7ffd4cde4010,  7 +        1f010, stack, 2, [I32 ]
[10] 7ffd4cde4854,  7 +        1f854, stack, 2, [I32 ]
[11] 7ffd4cde4910,  7 +        1f910, stack, 2, [I32 ]
12> exit
```

As you can see this actually deleted `0,2`, because you are reading the set, which contains `size_t` (int64) as `unsigned` (int32).
Endianness considered, this issued `delete 0; delete 1`, which then deleted 0 and 2 due to match shifting.

Please make the set pointer a `size_t`, as you fill it with `size_t`, there's no point in type erasure.
If we'll need a more general set, we'll see later.

If a `size_t` is used, the bug disappears.

Reply:
Yeah I figured that cast was going to cause trouble eventually, however the problem is the `nth_match` matching (hehe) parameter is `unsigned` which of course isn't necessarily the same as `size_t`.

Reply:
Relics from the 32bit time, when everything was simpler and longs were the same size as integers...
I also remembered you use a 32bit OS, so it's understandable you didn't catch it.

I'll add a treewide patch to change the matches indexing to `size_t` everywhere (since technically it could hold a whole program, hence the need for full object size).

Once I'm done, you can rebase and make the set.buf pointer a `size_t*`, this program has honestly too much type punning already, no need to introduce more.

Reply:
As promised: b79a55b .

Reply:
Excellent, pushed. 32bit will be relevant for a little longer, at least until our '07 laptops die completely ;-)